Progress for e2fsck

jtulak opened the pull request:
These two commits add a progress reporting capabilities for check/repair of ext2/3/4 fs. It utilizes the existing infrastructure including `bd_utils_init_prog_reporting()`. An example of usage is here: https://github.com/jtulak/fd-issue, but it is the same as for any other of the few operations that currently support progress report.

vpodzime commented:
@jtulak would you mind submitting a PR for https://github.com/storaged-project/api-examples with that example? I think it could be useful for other people too. Thanks!

vpodzime commented on src/plugins/fs/ext.c:
```diff
@@ -19,6 +19,7 @@
 
 #include <blockdev/utils.h>
 #include <check_deps.h>
+#include <regex.h>
```
Please use the [regular expressions implementation from GLib](https://developer.gnome.org/glib/2.54/glib-Perl-compatible-regular-expressions.html)

vpodzime commented on src/plugins/fs/ext.c:
```diff
@@ -59,6 +60,81 @@ static guint32 fs_mode_util[BD_FS_MODE_LAST+1] = {
+static gint8 compute_percents(guint8 pass_cur, guint8 pass_total, int val_cur, int val_total)
+{
+    int perc;
+    int one_pass;
```
Use GLib types here. And also in the parameters.

vpodzime commented on src/plugins/fs/ext.c:
```diff
@@ -59,6 +60,81 @@ static guint32 fs_mode_util[BD_FS_MODE_LAST+1] = {
+    /* A magic number 5, e2fsck has 5 stages, but this can't be read from the output in advance. */
+    *completion = filter_line_fsck(line, 5, (GError **)NULL);
+    if (*completion < 0)
+        return 0;
```
This should be ``FALSE``

vpodzime commented on src/plugins/fs/ext.c:
```diff
@@ -59,6 +60,81 @@ static guint32 fs_mode_util[BD_FS_MODE_LAST+1] = {
+}
+
+
+
```
Two blank lines should be enough here.

jtulak commented:
I updated the commits, so the issues should be fixed now. And I'm going to send the PR to the api examples too. :)

I did some other changes to this PR: I added one other commit and updated the last commit to make it easier to use (before, the user had to set -C1 argument for e2fsck manually, because ext.c plugin file didn't knew if we are reporting progress or not. I added a detection for this...

vojtechtrefny commented:
Jenkins, ok to test.

vojtechtrefny commented on src/utils/exec.h:
```diff
@@ -60,6 +60,7 @@ gint bd_utils_version_cmp (const gchar *ver_string1, const gchar *ver_string2, G
 gboolean bd_utils_check_util_version (const gchar *util, const gchar *version, const gchar *version_arg, const gchar *version_regexp, GError **error);
 
 gboolean bd_utils_init_prog_reporting (BDUtilsProgFunc new_prog_func, GError **error);
+gboolean bd_utils_prog_reporting_initialized();
```
Please add space between the function name and parentheses (and in the definition too). Thanks

jtulak commented:
Ok. Into which test I should add it, `utils_test.py`? I can't find progress reporting tested anywhere in `tests` directory: `grep -ri "prog_report" tests/` 

vojtechtrefny commented:
I think we don't have a test for progress reporting yet. The best place is probably `fs_test.py` -- you can add a new test case similar to `ExtTestCheck` just with the progress reporting enabled.

jtulak commented:
I pushed a new version with the space in the function header and tests. At the end, I made two of them: one to see if the progress reporting initialization works in Utils, and one in fs to see if ext4 check reports some progress.

vpodzime commented on src/utils/exec.c:
```diff
@@ -727,6 +727,17 @@ gboolean bd_utils_init_prog_reporting (BDUtilsProgFunc new_prog_func, GError **e
+gboolean bd_utils_prog_reporting_initialized () {
+    if (prog_func)
+        return TRUE;
+    return FALSE;
```
This can be reduced to ``return prog_func != NULL;``

vpodzime commented on src/plugins/fs/ext.c:
```diff
@@ -59,6 +59,75 @@ static guint32 fs_mode_util[BD_FS_MODE_LAST+1] = {
 #define UNUSED __attribute__((unused))
 
+static gint8 compute_percents(guint8 pass_cur, guint8 pass_total, gint val_cur, gint val_total)
+{
```
Put the brace to the end of the previous line, please.

vpodzime commented on src/plugins/fs/ext.c:
```diff
@@ -59,6 +59,75 @@ static guint32 fs_mode_util[BD_FS_MODE_LAST+1] = {
+ */
+static gint8 filter_line_fsck(const gchar * line, guint8 total_stages, GError **error) {
+    static GRegex *regex;
+    static int reginit = 0;
```
Please rename ``regex`` to something that suggests it's specific for this function. Also, either change ``reginit`` to ``gboolean`` or get rid of it altogether (use ``regex != NULL``).

vpodzime commented on src/plugins/fs/ext.c:
```diff
@@ -59,6 +59,75 @@ static guint32 fs_mode_util[BD_FS_MODE_LAST+1] = {
+    if (g_regex_match (regex, line, 0, &match_info)) {
+        guint8 stage;
+        long val_cur;
+        long val_total;
```
``g_ascii_strtoll()`` returns a ``gint64``. And I think ``g_ascii_strtoull()`` would be better here, right?

vpodzime commented on src/plugins/fs/ext.c:
```diff
@@ -59,6 +59,75 @@ static guint32 fs_mode_util[BD_FS_MODE_LAST+1] = {
+    /* A magic number 5, e2fsck has 5 stages, but this can't be read from the output in advance. */
+    gint8 perc;
+
+    perc = filter_line_fsck(line, 5, (GError **)NULL);
```
Please pass a real ``GError**`` here and use it above to report the error from ``g_regex_new()``

jtulak commented:
UPDATE sections removed and the other issues are fixed as well.

vpodzime commented on src/plugins/fs/ext.c:
```diff
@@ -59,6 +59,76 @@ static guint32 fs_mode_util[BD_FS_MODE_LAST+1] = {
+    if (g_regex_match (output_regex, line, 0, &match_info)) {
+        guint8 stage;
+        long val_cur;
+        long val_total;
```
These two should be ``gint64``

vpodzime commented on src/plugins/fs/ext.c:
```diff
@@ -59,6 +59,76 @@ static guint32 fs_mode_util[BD_FS_MODE_LAST+1] = {
 
 #define UNUSED __attribute__((unused))
 
+static gint8 compute_percents(guint8 pass_cur, guint8 pass_total, gint val_cur, gint val_total) {
```
Please add a space between function names an parentheses. In definitions as well as where the functions are called.

jtulak commented:
Hmmm, I'm not sure what is wrong with the last changes, why some Jenkins tests failed. Could the RPM build just get stuck?

vojtechtrefny commented:
> Hmmm, I'm not sure what is wrong with the last changes, why some Jenkins tests failed. Could the RPM build just get stuck?

I can ignore it, tests passed but Jenkins failed to copy the results to fedorapeople.

vpodzime commented on src/plugins/fs/ext.c:
```diff
@@ -59,6 +59,76 @@ static guint32 fs_mode_util[BD_FS_MODE_LAST+1] = {
+         */
+        stage = (guint8) g_ascii_strtoull(g_match_info_fetch (match_info, 1), (char **)NULL, 10);
+        val_cur = g_ascii_strtoll(g_match_info_fetch (match_info, 2), (char **)NULL, 10);
+        val_total = g_ascii_strtoll(g_match_info_fetch (match_info, 3), (char **)NULL, 10);
```
Still some missing spaces between function names and ``(``s in the above and also in a couple of places below. @jtulak, please fix this so that the PR can be merged.

jtulak replied:
OK, I tried to find any other case with a `[a-zA-Z0-9_](` regex in my files and it should be done.

vpodzime commented:
Thanks!
